File: Cpp/155.cpp
/*
Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

push(x) -- Push element x onto stack.
pop() -- Removes the element on top of the stack.
top() -- Get the top element.
getMin() -- Retrieve the minimum element in the stack.
Example:
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin();   --> Returns -3.
minStack.pop();
minStack.top();      --> Returns 0.
minStack.getMin();   --> Returns -2.
 * */
class MinStack {
private:
    stack<int> value, min_value;
public:
    /** initialize your data structure here. */
    MinStack() {
        
    }
    
    void push(int x) {
        value.push(x);
        if(min_value.empty() || min_value.top()>=x){
            min_value.push(x);
        }
    }
    
    void pop() {
        if(!min_value.empty() && min_value.top() == value.top()){
            min_value.pop();
        }
        value.pop();
    }
    
    int top() {
        return value.top();
    }
    
    int getMin() {
        return min_value.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Python/155.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.mstack = []
        

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        if self.stack == list() or x<=self.mstack[-1]:
            self.mstack.append(x)
        self.stack.append(x)
        

    def pop(self):
        """
        :rtype: None
        """
        if self.stack:
            if self.stack[-1] == self.mstack[-1]:
                self.mstack.pop()
            self.stack.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]
        

    def getMin(self):
        """
        :rtype: int
        """
        return self.mstack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
